File: materials/chat-client/python/chat_client.py
import asyncio
import json
import os
import sys

import websockets


SERVER_URL = os.environ.get("SERVER_URL", "ws://localhost:3000")


async def read_messages(ws):
    async for raw in ws:
        msg = json.loads(raw)
        match msg.get("type"):
            case "message":
                print(f"[{msg['name']}] {msg['text']}")
            case "join":
                print(f"--- {msg['name']} joined ---")
            case "leave":
                print(f"--- {msg['name']} left ---")
            case "error":
                print(f"Error: {msg['message']}")


async def send_messages(ws):
    loop = asyncio.get_event_loop()
    while True:
        text = await loop.run_in_executor(None, sys.stdin.readline)
        if not text:
            break
        text = text.strip()
        if text:
            await ws.send(json.dumps({"type": "message", "text": text}))


async def main():
    name = input("Enter your name: ").strip()
    if not name:
        print("Name cannot be empty.")
        return

    url = f"{SERVER_URL}?name={name}"

    try:
        async with websockets.connect(url) as ws:
            read_task = asyncio.create_task(read_messages(ws))
            send_task = asyncio.create_task(send_messages(ws))

            done, pending = await asyncio.wait(
                [read_task, send_task], return_when=asyncio.FIRST_COMPLETED
            )
            for task in pending:
                task.cancel()

            for task in done:
                if task.exception():
                    raise task.exception()

    except websockets.ConnectionClosed:
        print("\nConnection closed.")
    except (OSError, websockets.WebSocketException) as e:
        print(f"\nConnection error: {e}")


if __name__ == "__main__":
    asyncio.run(main())

File: materials/chat-client/node/chat_client.js
import readline from "readline";
import WebSocket from "ws";

const SERVER_URL = process.env.SERVER_URL || "ws://localhost:3000";

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

rl.question("What is your name? ", (name) => {
  const ws = new WebSocket(`${SERVER_URL}?name=${encodeURIComponent(name)}`);

  ws.on("open", () => {
    console.log(`Connected to ${SERVER_URL} as ${name}`);
    rl.on("line", (text) => {
      if (text.trim()) {
        ws.send(JSON.stringify({ type: "message", text }));
      }
    });
  });

  ws.on("message", (data) => {
    const msg = JSON.parse(data);
    switch (msg.type) {
      case "message":
        console.log(`[${msg.name}] ${msg.text}`);
        break;
      case "join":
        console.log(`--- ${msg.name} joined ---`);
        break;
      case "leave":
        console.log(`--- ${msg.name} left ---`);
        break;
      case "error":
        console.log(`Error: ${msg.message || msg.text}`);
        break;
    }
  });

  ws.on("close", () => {
    console.log("Disconnected from server.");
    process.exit(0);
  });

  ws.on("error", (err) => {
    console.error(`Connection error: ${err.message}`);
    process.exit(1);
  });
});
